File: gcode_viewer_v2/parser.py
"""G-code parser. Pure logic — no UI deps.

Ported from gcode_preview.py. The state machine handles modal G-code:
G0/G1/G2/G3 motion modes, G90/G91 distance modes, M3/M5 spindle, and per-axis
position carry-over (X/Y/Z hold their previous value when omitted).

Arcs (G2/G3) are linearized via arc_points() using I/J center-offset form.
R-form arcs are not currently supported (silently produce a degenerate move).
"""

import math
import re

SAFE_Z = 0.0
DEFAULT_FEED = 500.0
RAPID_FEED = 2500.0


class Move:
    """One toolpath segment (or arc, expanded into multiple points).

    kind: "G0" | "G1" | "G2" | "G3"
    sx, sy, sz: start position
    ex, ey, ez: end position
    feed: F value at the time of this move (mm/min)
    spindle: True if M3 active
    points: list of (x, y, z) tuples — for straight moves, [start, end];
            for arcs, multiple linearized points
    length: total path length in mm
    duration: elapsed time in seconds at the move's feed (rapids use RAPID_FEED)
    """

    def __init__(self, kind, sx, sy, sz, ex, ey, ez, feed, spindle, points=None):
        self.kind = kind
        self.sx = sx
        self.sy = sy
        self.sz = sz
        self.ex = ex
        self.ey = ey
        self.ez = ez
        self.feed = feed if feed > 0 else DEFAULT_FEED
        self.spindle = spindle
        self.points = points or [(sx, sy, sz), (ex, ey, ez)]

        self.length = 0.0
        for i in range(1, len(self.points)):
            self.length += math.dist(self.points[i - 1], self.points[i])

        used_feed = RAPID_FEED if kind == "G0" else self.feed
        self.duration = self.length / (used_feed / 60.0) if used_feed > 0 else 0.01


def clean(line):
    """Strip parenthesized comments and ;-comments, uppercase, and trim."""
    line = re.sub(r"\(.*?\)", "", line)
    line = line.split(";")[0]
    return line.strip().upper()


def val(line, letter):
    """Extract a numeric value following the given letter (e.g. 'X12.5').
    Returns None if the letter is not present.
    """
    m = re.search(letter + r"(-?\d+\.?\d*)", line)
    return float(m.group(1)) if m else None


def arc_points(sx, sy, sz, ex, ey, ez, i, j, clockwise):
    """Linearize a G2/G3 arc from (sx,sy) to (ex,ey) about center offsets (i,j).

    The center is at (sx+i, sy+j). Z linearly interpolates from sz to ez along
    the arc parameter. Step count scales with arc length to keep facets ~1.5mm.
    """
    cx = sx + i
    cy = sy + j
    r = math.hypot(sx - cx, sy - cy)

    a1 = math.atan2(sy - cy, sx - cx)
    a2 = math.atan2(ey - cy, ex - cx)

    if clockwise:
        if a2 > a1:
            a2 -= math.tau
    else:
        if a2 < a1:
            a2 += math.tau

    steps = max(16, int(abs(a2 - a1) * r / 1.5))
    pts = []
    for n in range(steps + 1):
        t = n / steps
        a = a1 + (a2 - a1) * t
        x = cx + math.cos(a) * r
        y = cy + math.sin(a) * r
        z = sz + (ez - sz) * t
        pts.append((x, y, z))
    return pts


def parse(text):
    """Parse a G-code program (string) into a list of Move objects.

    Modal state tracked: position (x/y/z), feed, spindle, distance mode
    (G90 abs / G91 rel), and current motion mode (G0/G1/G2/G3 — sticky
    across lines).

    Lines with no XYZ change are skipped (no Move emitted) — feed-only
    lines like "F500" update state but don't appear in the output.
    """
    x = y = z = 0.0
    feed = DEFAULT_FEED
    spindle = False
    absolute = True
    mode = "G0"

    moves = []

    for raw in text.splitlines():
        line = clean(raw)
        if not line:
            continue

        if "G90" in line:
            absolute = True
        if "G91" in line:
            absolute = False

        if "M3" in line or "M03" in line:
            spindle = True
        if "M5" in line or "M05" in line:
            spindle = False

        if "G0" in line or "G00" in line:
            mode = "G0"
        elif "G1" in line or "G01" in line:
            mode = "G1"
        elif "G2" in line or "G02" in line:
            mode = "G2"
        elif "G3" in line or "G03" in line:
            mode = "G3"

        nf = val(line, "F")
        if nf is not None:
            feed = nf

        nx = val(line, "X")
        ny = val(line, "Y")
        nz = val(line, "Z")

        if nx is None and ny is None and nz is None:
            continue

        sx, sy, sz = x, y, z
        ex = x if nx is None else (nx if absolute else x + nx)
        ey = y if ny is None else (ny if absolute else y + ny)
        ez = z if nz is None else (nz if absolute else z + nz)

        if mode in ("G2", "G3"):
            i = val(line, "I") or 0
            j = val(line, "J") or 0
            points = arc_points(sx, sy, sz, ex, ey, ez, i, j, mode == "G2")
        else:
            points = [(sx, sy, sz), (ex, ey, ez)]

        moves.append(Move(mode, sx, sy, sz, ex, ey, ez, feed, spindle, points))
        x, y, z = ex, ey, ez

    return moves


def parse_file(path):
    """Convenience wrapper: read a .nc / .gcode file from disk and parse it."""
    with open(path, "r", errors="ignore") as f:
        return parse(f.read())


def bounds(moves):
    """Return ((min_x, min_y, min_z), (max_x, max_y, max_z)) over all move points,
    or None if the moves list is empty.
    """
    if not moves:
        return None
    xs = []
    ys = []
    zs = []
    for m in moves:
        for px, py, pz in m.points:
            xs.append(px)
            ys.append(py)
            zs.append(pz)
    return (min(xs), min(ys), min(zs)), (max(xs), max(ys), max(zs))


def deepest_cut_z(moves):
    """Return the most-negative Z reached during any cutting (non-G0) move,
    or -1.0 as a fallback if no cut goes below zero.
    """
    deepest = 0.0
    for m in moves:
        if m.kind == "G0":
            continue
        for p in m.points:
            if p[2] < deepest:
                deepest = p[2]
    return deepest if deepest < 0 else -1.0

File: gcode_viewer_v3/internal/parser/parser.go
// Package parser implements a G-code parser. Pure logic — no UI deps.
//
// Ported 1:1 from gcode_viewer_v2/parser.py. Handles modal G-code:
// G0/G1/G2/G3 motion modes, G90/G91 distance modes, M3/M5 spindle, and
// per-axis position carry-over (X/Y/Z hold their previous value when
// omitted). Arcs (G2/G3) are linearized via I/J center-offset form.
package parser

import (
	"math"
	"os"
	"regexp"
	"strconv"
	"strings"
)

const (
	SafeZ       = 0.0
	DefaultFeed = 500.0
	RapidFeed   = 2500.0
)

// Point is a single 3D coordinate.
type Point struct {
	X, Y, Z float64
}

// Move is one toolpath segment (or arc, expanded into multiple points).
type Move struct {
	Kind       string
	Sx, Sy, Sz float64
	Ex, Ey, Ez float64
	Feed       float64
	Spindle    bool
	Points     []Point
	Length     float64
	Duration   float64
}

func newMove(kind string, sx, sy, sz, ex, ey, ez, feed float64, spindle bool, points []Point) *Move {
	if feed <= 0 {
		feed = DefaultFeed
	}
	if points == nil {
		points = []Point{{sx, sy, sz}, {ex, ey, ez}}
	}
	m := &Move{
		Kind: kind, Sx: sx, Sy: sy, Sz: sz, Ex: ex, Ey: ey, Ez: ez,
		Feed: feed, Spindle: spindle, Points: points,
	}
	for i := 1; i < len(points); i++ {
		m.Length += dist3(points[i-1], points[i])
	}
	usedFeed := feed
	if kind == "G0" {
		usedFeed = RapidFeed
	}
	if usedFeed > 0 {
		m.Duration = m.Length / (usedFeed / 60.0)
	} else {
		m.Duration = 0.01
	}
	return m
}

func dist3(a, b Point) float64 {
	dx := b.X - a.X
	dy := b.Y - a.Y
	dz := b.Z - a.Z
	return math.Sqrt(dx*dx + dy*dy + dz*dz)
}

var (
	reComment = regexp.MustCompile(`\(.*?\)`)

	// reGWord matches every G-word on a line so motionMode can compare whole
	// numbers instead of substrings.
	reGWord = regexp.MustCompile(`G(\d+)`)

	reLetterVal = map[byte]*regexp.Regexp{
		'X': regexp.MustCompile(`X(-?\d+\.?\d*)`),
		'Y': regexp.MustCompile(`Y(-?\d+\.?\d*)`),
		'Z': regexp.MustCompile(`Z(-?\d+\.?\d*)`),
		'I': regexp.MustCompile(`I(-?\d+\.?\d*)`),
		'J': regexp.MustCompile(`J(-?\d+\.?\d*)`),
		'F': regexp.MustCompile(`F(-?\d+\.?\d*)`),
	}
)

func clean(line string) string {
	line = reComment.ReplaceAllString(line, "")
	if i := strings.Index(line, ";"); i >= 0 {
		line = line[:i]
	}
	return strings.ToUpper(strings.TrimSpace(line))
}

// motionMode returns the motion mode named by the first G-word on the line
// that is one of G0/G1/G2/G3, with leading zeros allowed (so "G01" is G1 and
// "G03" is G3). ok is false when the line names no motion word at all, in
// which case the caller leaves the sticky mode alone.
//
// This deliberately compares whole numbers rather than substrings. The
// obvious strings.Contains(line, "G0") test — which this port and the Python
// reference both used originally — misfires on any G-word that merely
// contains those two characters:
//
//	"G01 X10"     contains "G0" → a linear cut was read as a rapid
//	"G03 X.. I.." contains "G0" → a helical arc was read as a rapid, so it
//	              neither linearized nor removed any material
//	"G90 G21 G17" contains "G1" (inside G17) → left the parser in G1 mode
//	"G21"         contains "G2" → left the parser in ARC mode, so the next
//	              bare coordinate line would be linearized as an arc
//
// Bare-notation files never tripped the first two, which is why the old
// behaviour survived; HoleGen output hits all of them.
func motionMode(line string) (string, bool) {
	for _, m := range reGWord.FindAllStringSubmatch(line, -1) {
		n, err := strconv.Atoi(m[1])
		if err != nil {
			continue // absurdly long digit run — not a motion word
		}
		if n <= 3 {
			return "G" + strconv.Itoa(n), true
		}
	}
	return "", false
}

// val extracts the numeric value following the given letter.
// Returns (value, true) if found, (0, false) otherwise.
func val(line string, letter byte) (float64, bool) {
	re := reLetterVal[letter]
	m := re.FindStringSubmatch(line)
	if m == nil {
		return 0, false
	}
	f, err := strconv.ParseFloat(m[1], 64)
	if err != nil {
		return 0, false
	}
	return f, true
}

// Parse parses a G-code program (string) into Move objects.
//
// Modal state tracked: position (x/y/z), feed, spindle, distance mode (G90
// abs / G91 rel), and current motion mode (G0/G1/G2/G3 — sticky across lines).
//
// Lines with no XYZ change are skipped — feed-only lines like "F500" update
// state but don't appear in the output.
func Parse(text string) []*Move {
	var x, y, z float64
	feed := float64(DefaultFeed)
	spindle := false
	absolute := true
	mode := "G0"

	var moves []*Move

	for _, raw := range strings.Split(text, "\n") {
		line := clean(raw)
		if line == "" {
			continue
		}

		if strings.Contains(line, "G90") {
			absolute = true
		}
		if strings.Contains(line, "G91") {
			absolute = false
		}

		if strings.Contains(line, "M3") || strings.Contains(line, "M03") {
			spindle = true
		}
		if strings.Contains(line, "M5") || strings.Contains(line, "M05") {
			spindle = false
		}

		// Motion mode is sticky: a line with coordinates but no G-word keeps
		// whatever mode was last set. gcode_viewer_v2/parser.py carries the
		// same whole-number matching.
		if mm, ok := motionMode(line); ok {
			mode = mm
		}

		if nf, ok := val(line, 'F'); ok {
			feed = nf
		}

		nx, hx := val(line, 'X')
		ny, hy := val(line, 'Y')
		nz, hz := val(line, 'Z')
		if !hx && !hy && !hz {
			continue
		}

		sx, sy, sz := x, y, z
		ex := x
		if hx {
			if absolute {
				ex = nx
			} else {
				ex = x + nx
			}
		}
		ey := y
		if hy {
			if absolute {
				ey = ny
			} else {
				ey = y + ny
			}
		}
		ez := z
		if hz {
			if absolute {
				ez = nz
			} else {
				ez = z + nz
			}
		}

		var points []Point
		if mode == "G2" || mode == "G3" {
			i, _ := val(line, 'I')
			j, _ := val(line, 'J')
			points = ArcPoints(sx, sy, sz, ex, ey, ez, i, j, mode == "G2")
		} else {
			points = []Point{{sx, sy, sz}, {ex, ey, ez}}
		}

		moves = append(moves, newMove(mode, sx, sy, sz, ex, ey, ez, feed, spindle, points))
		x, y, z = ex, ey, ez
	}

	return moves
}

// ParseFile reads a .nc / .gcode file from disk and parses it.
func ParseFile(path string) ([]*Move, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return Parse(string(data)), nil
}

// Bounds returns ((min_x, min_y, min_z), (max_x, max_y, max_z)) over all
// move points. Returns ok=false if the moves slice is empty.
func Bounds(moves []*Move) (min, max Point, ok bool) {
	if len(moves) == 0 {
		return Point{}, Point{}, false
	}
	min = moves[0].Points[0]
	max = min
	for _, m := range moves {
		for _, p := range m.Points {
			if p.X < min.X {
				min.X = p.X
			}
			if p.Y < min.Y {
				min.Y = p.Y
			}
			if p.Z < min.Z {
				min.Z = p.Z
			}
			if p.X > max.X {
				max.X = p.X
			}
			if p.Y > max.Y {
				max.Y = p.Y
			}
			if p.Z > max.Z {
				max.Z = p.Z
			}
		}
	}
	return min, max, true
}

// DeepestCutZ returns the most-negative Z reached during any cutting (non-G0)
// move, or -1.0 as a fallback if no cut goes below zero.
func DeepestCutZ(moves []*Move) float64 {
	deepest := 0.0
	for _, m := range moves {
		if m.Kind == "G0" {
			continue
		}
		for _, p := range m.Points {
			if p.Z < deepest {
				deepest = p.Z
			}
		}
	}
	if deepest >= 0 {
		return -1.0
	}
	return deepest
}
